packages/zoe/src/contracts/multipoolAutoswap/getCurrentPrice.js

```diff
@@ -55,30 +64,129 @@ export const makeGetCurrentPrice = (isSecondary, isCentral, getPool) => {
    * @returns {Amount} The amount required to be paid in order to gain amountOut
    */
   const getInputForGivenOutput = (amountOut, brandIn) => {
-    const { brand: brandOut, value: outputValue } = amountOut;
+    const { brand: brandOut } = amountOut;
 
     if (isCentral(brandIn) && isSecondary(brandOut)) {
-      return getPool(brandOut).getCentralToSecondaryOutputPrice(outputValue);
+      return getPool(brandOut).getPriceGivenRequiredOutput(brandIn, amountOut)
+        .amountIn;
     }
 
     if (isSecondary(brandIn) && isCentral(brandOut)) {
-      return getPool(brandIn).getSecondaryToCentralOutputPrice(outputValue);
+      return getPool(brandIn).getPriceGivenRequiredOutput(brandIn, amountOut)
+        .amountIn;
     }
 
     if (isSecondary(brandIn) && isSecondary(brandOut)) {
       // We must do two consecutive calls to get the price: from
       // the brandIn to the central token, then from the central
       // token to the brandOut.
-      const centralTokenAmount = getPool(
+      const { amountIn: centralTokenAmount } = getPool(
+        brandIn,
+      ).getPriceGivenRequiredOutput(brandIn, amountOut);
+      return getPool(brandOut).getPriceGivenRequiredOutput(
+        centralBrand,
+        centralTokenAmount,
+      ).amountIn;
+    }
+
+    throw new Error(`brands were not recognized`);
+  };
+
+  const getPriceGivenAvailableInput = (amountIn, brandOut) => {
+    const { brand: brandIn } = amountIn;
+
+    if (isCentral(brandIn) && isSecondary(brandOut)) {
+      return getPool(brandOut).getPriceGivenAvailableInput(amountIn, brandOut);
+    } else if (isSecondary(brandIn) && isCentral(brandOut)) {
+      return getPool(brandIn).getPriceGivenAvailableInput(amountIn, brandOut);
+    } else if (isSecondary(brandIn) && isSecondary(brandOut)) {
+      // We must do two consecutive getPriceGivenAvailableInput() calls,
+      // followed by a call to getPriceGivenRequiredOutput().
+      // 1) from amountIn to the central token, which tells us how much central
+      // would be provided for amountIn (centralAmount)
+      // 2) from centralAmount to brandOut, which tells us how much of brandOut
+      // will be provided (amountOut) as well as the minimum price in central
+      // tokens (reducedCentralAmount), then finally
+      // 3) call getPriceGivenRequiredOutput() to see if the same proceeds can
+      // be purchased for less (reducedAmountIn).
+
+      const brandInPool = getPool(brandIn);
+      const brandOutPool = getPool(brandOut);
+      const {
+        amountOut: centralAmount,
+      } = brandInPool.getPriceGivenAvailableInput(amountIn, centralBrand);
+      const {
+        amountIn: reducedCentralAmount,
+        amountOut,
+      } = brandOutPool.getPriceGivenAvailableInput(centralAmount, brandOut);
+
```

Comment:
All of this code is the same as `getOutputForGivenInput`, except for whether just `amountOut` is returned rather than a record. Given how easy it is to get this logic wrong, I think we don't want to duplicate it, but rather reuse it.

Reply:
I think we should retain the versions that return just an amount, as it matches users' expectations coming from Uniswap (https://github.com/Uniswap/uniswap-v1/blob/master/contracts/uniswap_exchange.vy)

Reply:
Just to confirm, do you mean retain the surface API, with it calling the other method?